### PR TITLE
Direct passing WebIdentity token to AWS provider

### PR DESCRIPTION
**Pull request by ed-shilo**
## Change Overview

Allowing to pass Role ARN and Web Identity token for token-based AWS authentication. 

## Pull request type

Please check the type of change your PR introduces:
- [ ] :construction: Work in Progress
- [ ] :rainbow: Refactoring (no functional changes, no api changes)
- [ ] :hamster: Trivial/Minor
- [ ] :bug: Bugfix
- [x] :sunflower: Feature
- [ ] :world_map: Documentation
- [ ] :robot: Test

## Test Plan

- [x] :muscle: Manual
- [ ] :zap: Unit test
- [ ] :green_heart: E2E

### CODE REVIEW

File: pkg/aws/aws.go [hunk 1]
```diff
@@ -76,15 +88,24 @@ func authenticateAWSCredentials(ctx context.Context, config map[string]string, a
...
+		tokenFetcher := stscreds.FetchTokenPath(os.Getenv(webIdentityTokenFilePathEnvKey))
+		var err error
+		if creds, err = getCredentialsWithDuration(os.Getenv(roleARNEnvKey), tokenFetcher, assumeRoleDuration); err != nil {
+			return nil, "", nil
 		}
-		creds = getCredentialsWithDuration(sess, assumeRoleDuration)
 		assumedRole = os.Getenv(roleARNEnvKey)
 	default:
-		return nil, "", errors.New("AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY required to initialize AWS credentials")
+		return nil, "", errors.New(fmt.Sprintf("%s and %s required to initialize AWS credentials", AccessKeyID, SecretAccessKey))
```
Comment (viveksinghggits) [inline pkg/aws/aws.go:108]: Should we just use `fmt.Errorf` here?

**Comment by viveksinghggits** [comment]
LGTM overall.

File: pkg/aws/aws.go (hunk 1 above)
Comment (ed-shilo) [reply pkg/aws/aws.go]: Fixed, thanks

File: pkg/aws/aws.go [hunk 2]
```diff
@@ -60,6 +63,14 @@ const (
 	AssumeRoleDuration        = "assumeRoleDuration"
 )
 
+var _ stscreds.TokenFetcher = (*staticToken)(nil)
+
+type staticToken string
+
+func (f staticToken) FetchToken(ctx credentials.Context) ([]byte, error) {
```
Comment (leuyentran) [inline pkg/aws/aws.go:70]: [nit] some comments will help w understanding: Implements `TokenFetcher` interface from ... for static web identity token

File: pkg/aws/aws.go [hunk 3]
```diff
@@ -76,15 +87,24 @@ func authenticateAWSCredentials(ctx context.Context, config map[string]string, a
...
+		tokenFetcher := stscreds.FetchTokenPath(os.Getenv(webIdentityTokenFilePathEnvKey))
+		var err error
+		if creds, err = getCredentialsWithDuration(os.Getenv(roleARNEnvKey), tokenFetcher, assumeRoleDuration); err != nil {
+			return nil, "", nil
 		}
-		creds = getCredentialsWithDuration(sess, assumeRoleDuration)
 		assumedRole = os.Getenv(roleARNEnvKey)
 	default:
-		return nil, "", errors.New("AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY required to initialize AWS credentials")
+		return nil, "", errors.Errorf("%s and %s required to initialize AWS credentials", AccessKeyID, SecretAccessKey)
```
Comment (leuyentran) [inline pkg/aws/aws.go:107]:
```suggestion
		return nil, "", errors.Errorf("Missing AWS credentials, please check that either AWS access keys or web identity token are provided" )
```

File: pkg/aws/aws.go (hunk 2 above)
Comment (ed-shilo) [reply pkg/aws/aws.go:73]: Fixed

File: pkg/aws/aws.go [hunk 4]
```diff
@@ -76,15 +87,24 @@ func authenticateAWSCredentials(ctx context.Context, config map[string]string, a
 	case config[AccessKeyID] != "" && config[SecretAccessKey] != "":
 		// If AccessKeys were provided - use those
 		creds = credentials.NewStaticCredentials(config[AccessKeyID], config[SecretAccessKey], "")
+	case config[ConfigWebIdentityToken] != "" && config[ConfigRole] != "":
+		// If Web Identity token and role were provided - use them
```
[the remark this answers is not in the document]
Comment (ed-shilo) [reply pkg/aws/aws.go]: Tried to do my best to refactor this :)

File: pkg/aws/aws.go (hunk 3 above)
Comment (ed-shilo) [reply pkg/aws/aws.go]: Fixed

File: pkg/aws/aws.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (leuyentran) [reply pkg/aws/aws.go]: Thank you! It's much easier to read :pray:

I still think the function is doing too much, but it's hard to refactor without ripping off a lot of things given the current codebase. So for now this is good to me.

**Review by pavannd1 — APPROVED** [review]

File: pkg/aws/aws.go [hunk 5]
```diff
@@ -68,25 +82,77 @@ func durationFromString(config map[string]string) (time.Duration, error) {
...
+	if creds != nil {
 		assumedRole = os.Getenv(roleARNEnvKey)
-	default:
-		return nil, "", errors.New("AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY required to initialize AWS credentials")
+		return
 	}
-	return creds, assumedRole, nil
+
+	err = errors.New("Missing AWS credentials, please check that either AWS access keys or web identity token are provided")
+	return
```
Comment (denisvmedia) [inline pkg/aws/aws.go:113]: Why does this function uses naked return approach, while other functions use explicit returns?

Comment (ed-shilo) [reply pkg/aws/aws.go]: Changed to use explicit returns, but in this particular case using naked returns looks more laconic.

**Pull request merged**